libs/postgres_ffi/src/xlog_utils.rs

```diff
@@ -622,26 +621,50 @@ mod tests {
         }
         cfg.initdb().unwrap();
         let srv = cfg.start_server().unwrap();
-        let expected_wal_end: Lsn =
-            u64::from(C::craft(&mut srv.connect_with_timeout().unwrap()).unwrap()).into();
+        let (intermediate_lsns, expected_end_of_wal_partial): (Vec<Lsn>, Lsn) =
+            match C::craft(&mut srv.connect_with_timeout().unwrap()).unwrap() {
+                (a, b) => (
+                    a.iter().map(|&lsn| u64::from(lsn).into()).collect(),
+                    u64::from(b).into(),
+                ),
+            };
         srv.kill();
 
-        // 2. Pick WAL generated by initdb
-        let wal_dir = cfg.datadir.join("pg_wal");
-        let wal_seg_size = 16 * 1024 * 1024;
+        let get_last_segment = || {
+            let wal_dir = cfg.datadir.join("pg_wal");
+            let SEGMENT_NAME_REGEX = Regex::new(r"^[0-9A-F]{24}$").unwrap();
+            wal_dir
+                .read_dir()
+                .unwrap()
+                .map(|f| f.unwrap().file_name().into_string().unwrap())
+                .filter(|fname| SEGMENT_NAME_REGEX.is_match(fname))
+                .max()
+                .unwrap()
+        };
 
-        // 3. Check end_of_wal on non-partial WAL segment (we treat it as fully populated)
-        let (wal_end, tli) = find_end_of_wal(&wal_dir, wal_seg_size, true, Lsn(0)).unwrap();
-        let wal_end = Lsn(wal_end);
-        info!(
-            "find_end_of_wal returned (wal_end={}, tli={})",
-            wal_end, tli
-        );
-        assert_eq!(wal_end, expected_end_of_wal_non_partial);
+        // Check find_end_of_wal on the initial WAL
+        let last_segment = get_last_segment();
+        check_pg_waldump_end_of_wal(&cfg, &last_segment, expected_end_of_wal_partial);
+        for start_lsn in std::iter::once(Lsn(0)).chain(intermediate_lsns) {
+            // TODO: zero WAL files before start_lsn
```

Comment:
It is to make sure we're really reading from `start_lsn`, and not just from the start every time.